features/202-session-management/design.md

```diff
@@ -0,0 +1,215 @@
+
+APIs would be provided to manage sessions of a user.
+
+Session list API returns a list of valid session of a user. Each list entry
+contains the session ID and session attributes. Access token / refresh token of
+the session would not be included. Parsed user agent data is also returned.
+
+Revoke sessions API revokes a specific session, or all sessions other than the
+requesting session. A `session_delete` event would be generated for all sessions
+to be revoked.
+
+Update session API allows updating name of session. Custom attributes can also
+be updated using master key.
+
+When the name of session, or custom attributes are updated, a `session_update`
+event would be generated. Web-hook event context would include session
+attributes and mutations can be used to update them.
+
+
+## Access / Refresh Token
+
+Each session would have an access token, and optionally a refresh token. Both
+type of tokens are opaque: token formats are unspecified and developer should
+not attempt to interpret the content of tokens.
+
+Access token would always have a lifetime; access token would be treated as
+invalid after its expiry.
+
+Refresh token can be used to obtain a new access token. If the old access token
+is still valid, the old access token is invalidated; there would be at most one
+valid access token for a session at any time.
+
+Refresh token would always have a maximum lifetime.
+
+Optionally, a session idle timeout can be specified: refresh token (or access
+token if refresh token is disabled) must be used at least once before the
+timeout, otherwise the session would be expired.
+
+A session is invalidated if its identity is deleted, or its user is disabled.
+A session is expired if its refresh token is expired, or its access token is
+expired (if refresh token is disabled).
+If a session is invalidated, expired, revoked, or logout, its associated access
+token / refresh token would be treated as invalid.
+
+Client-side SDK should handle refreshing of expired access token automatically
+if refresh token is available; SDKs would not expose concept of access token /
+refresh token; however, developers should still handle situation where session
+is no longer valid due to various reasons.
+
+
+## Session Cookie
+
+Optionally, developer can choose to use cookies to transport session tokens, in
+order to facilitate server-side rendering.
+
+If developer choose to enable session cookie:
+- Refresh token would not be issued. (See [Appendix](#session-cookie-and-refresh-token))
+- Access token would be transported in session cookie.
+- Access token would not be returned from APIs.
+- Gateway would read access token from cookie only, ignoring `Authorization`
+  header. If a token is present in both header and cookie, it would be treated
+  as if no token is present.
+- The session cookie would be cleared when:
+    - access token will be returned when header transport is used; or
+    - token is present in both header and cookie; or
```

Comment:
Would it be better to present an error instead of fail silently by clear session cookie in this case?

Reply:
- We do not know how to present the error in a correct format: it may be JSON for API, and HTML for rendered pages.
- This case can occurred only due to configuration changes, which is considered an edge case.

Reply:
What about just an HTTP Error without content? Sounds like could be 400. As you said it is edge case, so a HTTP Error with a simple text seems sufficient to me.

Or it is intended to fail silently, so when a configuration changes, next request will go through because the cookies are cleared?

Reply:
(If it is the later case I think we should include the "Why" in the spec, it is an important piece of imformation why we intend to silently handle an error.

Reply:
Strictly speaking, it is not an error or failure: it is just treated as session not found. If session is not found (e.g. revoked), the session cookie is cleared.

Reply:
But it is not expected, right? In that case, I think we can treat it as an error.

Again, or unless it is intended to fail silently? In that case, can we write down the reason in spec?